Refactor filterChooser

lbwexler opened the pull request:
+ Add validation of favorites, value
+ Clarify structure of options
+ Breakout Query processing in to its own file.
+ Tweak to option sorting logic to support future changes

Hoist P/R Checklist
-------------------

**Pull request authors:** Review and check off the below. Items that do not apply can also be
checked off to indicate they have been considered. If unclear if a step is relevant, please leave
unchecked and note in comments.

- [x] Caught up with `develop` branch as of last change.
- [x] Added CHANGELOG entry, or determined not required.
- [x] Reviewed for breaking changes, added `breaking-change` label + CHANGELOG if so.
- [x] Updated doc comments / prop-types, or determined not required.
- [x] Reviewed and tested on Mobile, or determined not required.
- [x] Created Toolbox branch / PR, or determined not required.

**If your change is still a WIP**, please use the "Create draft pull request" option in the split
button below to indicate it is not ready yet for a final review.

> **Pull request reviewers:** when merging this P/R, please consider using a *squash commit* to
> collapse multiple intermediate commits into a single commit representing the overall feature
> change. This helps keep the commit log clean and easy to scan across releases. PRs containing a
> single commit should be *rebased* when possible.

TomTirapani commented on cmp/filter/FilterChooserFieldSpec.js:
```diff
@@ -51,7 +48,64 @@ export class FilterChooserFieldSpec {
+        this.store = store;
+        const storeField = this.storeField = store?.getField(field);
+
+        this.fieldType = this.fieldType ?? storeField?.type ?? FieldType.AUTO;
```
Shouldn't this be `this.fieldType = fieldType`?

amcclain commented on cmp/filter/FilterChooserFieldSpec.js:
```diff
@@ -51,7 +50,66 @@ export class FilterChooserFieldSpec {
+     *      components to aid usability.
+     * @param {Store} [store] - set from controlling `FilterChooserModel.sourceStore` config, used
+     *      to source matching data `Field` and extract values if configured.
+     */
```
I had specifically documented these via a typedef so we could then use it in the documentation for `FilterChooserModel.fieldSpecs` - the actual entry point for the developer. I had done this for `FieldConfig` as well (in Field.js).

IMO pushing it back to the ctor here just encourages a dev to construct this object, which we don't want them to do, and doesn't provide them any inline help when specifying fieldSpecs, which is exactly what we _do_ need them to do.

I left this in its new location on the ctor and updated the tags to `@param` consistently but would like to discuss as this is a pattern we have in a few places where we have really nice docs, but they aren't easily discoverable and don't provide the dev-time assistance they could because they are on a ctor that devs never call (`Column` is another prime example).

lbwexler commented:
Happy to discuss this of course  I have a huge number of additional changes
in my sandbox from last night.  Will look to get in for review ASAP
tommorrow.

I think we do need

On Sat, Aug 29, 2020, 6:57 PM Anselm McClain <notifications@github.com>
wrote:

> *@amcclain* commented on this pull request.
>
> Hi - I know this is a WIP so didn't get too much into looking @ the new
> behavior of the querying or worrying about cleanups on FilterChooserModel.
> Did make some incremental doc changes (hopefully correct) to help keep
> things straight in my own mind as they evolve. Look forward to discussing
> more on Monday!
> ------------------------------
>
> In admin/tabs/activity/tracking/ActivityTrackingModel.js
> <https://github.com/xh/hoist-react/pull/2047#discussion_r479693822>:
>
> > @@ -107,11 +107,11 @@ export class ActivityTrackingModel {
>                  'browser',
>                  {
>                      field: 'msg',
> -                    suggestValues: false
> +                    suggestValues: true
>
> For testing or something we think we would want?
> ------------------------------
>
> In admin/tabs/activity/tracking/ActivityTrackingModel.js
> <https://github.com/xh/hoist-react/pull/2047#discussion_r479694012>:
>
> > @@ -98,7 +98,7 @@ export class ActivityTrackingModel {
>
>          this.filterChooserModel = new FilterChooserModel({
>              initialValue: this._defaultFilter,
> -            store: this.cube.store,
> +            sourceStore: this.cube.store,
>
> This is a nice example of the usefulness of a distinct sourceStore config
> ------------------------------
>
> In cmp/filter/FilterChooserFieldSpec.js
> <https://github.com/xh/hoist-react/pull/2047#discussion_r479697788>:
>
> > +     *      based on `fieldType`.
> +     * @param {boolean} [suggestValues] - true to provide auto-complete options with data values
> +     *      sourced either automatically from Store data or provided directly via the `values`
> +     *      config below. Default is based on fieldType. Set to `false` to disable extraction and
> +     *      suggestion of values from Store.
> +     * @param {[]} [values] - explicit list of available values to autocomplete for this field.
> +     *      If unspecified, values will be extracted and updated from Store data if applicable.
> +     * @param {FilterOptionValueRendererCb} [valueRenderer] - function to produce a suitably
> +     *      formatted string for display to the user for any given field value.
> +     * @param {FilterOptionValueParserCb} [valueParser] - function to parse user's input from a
> +     *      filter chooser control into a typed data value for use in filtering comparisons.
> +     * @param {*} [exampleValue] - sample / representative value displayed by `FilterChooser`
> +     *      components to aid usability.
> +     * @param {Store} [store] - set from controlling `FilterChooserModel.sourceStore` config, used
> +     *      to source matching data `Field` and extract values if configured.
> +     */
>
> I had specifically documented these via a typedef so we could then use it
> in the documentation for FilterChooserModel.fieldSpecs - the actual entry
> point for the developer. I had done this for FieldConfig as well (in
> Field.js).
>
> IMO pushing it back to the ctor here just encourages a dev to construct
> this object, which we don't want them to do, and doesn't provide them any
> inline help when specifying fieldSpecs, which is exactly what we *do*
> need them to do.
>
> I left this in its new location on the ctor and updated the tags to @param
> consistently but would like to discuss as this is a pattern we have in a
> few places where we have really nice docs, but they aren't easily
> discoverable and don't provide the dev-time assistance they could because
> they are on a ctor that devs never call (Column is another prime example).
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/xh/hoist-react/pull/2047#pullrequestreview-478157045>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AARTL2J4NDR5QDP6Y2QTUITSDGBVFANCNFSM4QNC3K4Q>
> .
>

lbwexler commented:
Have reviewed this extensively with Tom and John.  It's by no means perfect yet, but this moves the ball substantially forward from what we have, and the new structure has much more ability to take further enhancements.  

Am going to merge it it now, so that John can begin testing and we can show Dima.
